File: resources/scripts/components/transform.lua
ffi.cdef[[
typedef struct transform_component_t
{
  kmVec3 position;
  kmQuaternion rotation;
  kmVec3 scale;
  UUID parent;
  UUID firstChild;
  UUID nextSibling;
  bool dirty;
  kmVec3 globalPosition;
  kmQuaternion globalRotation;
  kmVec3 globalScale;
  kmVec3 lastGlobalPosition;
  kmQuaternion lastGlobalRotation;
  kmVec3 lastGlobalScale;
} TransformComponent;
]]

io.write("Defined Transform component for FFI\n")

local component = engine.components:register("transform", "TransformComponent")

local emptyID = engine.C.idFromName("")

function component:markDirty(scene)
  self.dirty = true

  if ffi.C.strcmp(self.firstChild.string, emptyID.string) ~= 0 then
    local child = scene:getComponent("transform", self.firstChild)
    child:markDirty(scene)
    while ffi.C.strcmp(child.nextSibling.string, emptyID.string) ~= 0 do
      child = scene:getComponent("transform", child.nextSibling);
      child:markDirty(scene);
    end
  end
end

function component:parent(scene, child, parent)
  local parentTransform = scene:getComponent("transform", parent)

  self.nextSibling = parentTransform.firstChild
  parentTransform.firstChild = child

  self.parent = parent

  self:markDirty(scene, child)
end

io.write("Registered Transform component\n")

File: src/game/components/transform.c
#include "components/transform.h"

void tMarkDirty(Scene *scene, UUID entityID)
{
	UUID transformComponentID = idFromName("transform");

	TransformComponent *trans =
		sceneGetComponentFromEntity(scene, entityID, transformComponentID);

	if (trans)
	{
		trans->dirty = true;

		TransformComponent *child = 0;

		for (UUID currentChild = trans->firstChild;
			 strcmp(currentChild.string, "");
			 currentChild = child->nextSibling)
		{
			child = sceneGetComponentFromEntity(
				scene,
				currentChild,
				transformComponentID);

			tMarkDirty(scene, currentChild);
		}
	}
}

void tDecomposeMat4(
	kmMat4 const *transform,
	kmVec3 *position,
	kmQuaternion *rotation,
	kmVec3 *scale)
{
	// TODO: make a way to get the stuff out of the matrix
}

kmMat4 tComposeMat4(
	kmVec3 const *position,
	kmQuaternion const *rotation,
	kmVec3 const *scale)
{
	kmQuaternion rotationQuat;
	kmQuaternionAssign(&rotationQuat, rotation);
	kmQuaternionNormalize(
		&rotationQuat,
		&rotationQuat);

	kmVec3 pos;
	kmVec3Assign(&pos, position);

	kmVec3 s;
	kmVec3Assign(&s, scale);

	kmMat4 rotationMatrix;
	kmMat4RotationQuaternion(&rotationMatrix, &rotationQuat);

	kmMat4 positionMatrix;
	kmMat4Translation(
		&positionMatrix,
		pos.x,
		pos.y,
		pos.z);

	kmMat4 scaleMatrix;
	kmMat4Scaling(
		&scaleMatrix,
		s.x,
		s.y,
		s.z);

	kmMat4 worldMatrix;
	kmMat4Multiply(
		&worldMatrix,
		&positionMatrix,
		kmMat4Multiply(&worldMatrix, &rotationMatrix, &scaleMatrix));

	return worldMatrix;
}
